### PR TITLE
Rework Pkmn Filter Functions

### PR DESCRIPTION
**Pull request by TurtIeSocks**
- Cleaned up code from my early PRs
- Fixed preset poke filter buttons, apparently they've been broken? 
- Removed generations file in favor of a simple check.
- Consolidated poke filter functions to reduce redundant code. I believe the new functions are also a little more robust for adding future filter options in the future
- Add kalos filters (must be coming soon, no?)
- Fix legendary/mythical in masterfile

### CODE REVIEW

File: static/js/index.js [hunk 1]
```diff
@@ -5579,85 +5576,62 @@ function getPokemonBestRank(greatLeague, ultraLeague) {
...
-    return pokemonGenerationDb.hoenn.includes(pokemonId);
-}
+let rarityFilter = (pokemonId, rarityTier) => pokemonRarity[rarityTier].includes(parseInt(pokemonId));
 
-function isSinnohPokemon(pokemonId) {
-    return pokemonGenerationDb.sinnoh.includes(pokemonId);
+let genFilter = (pokemonId, gen) => {
+    pokemonId = parseInt(pokemonId);
+    let match = pokemonId >= gen[0] && pokemonId <= gen[1] ? true : false;
+    return match;
```
Comment (Mygod) [inline static/js/index.js:5586]: What about regional forms?

Comment (TurtIeSocks) [reply static/js/index.js:5586]: See next function?

**Comment by versx** [comment]
Looks good, should we make the generation ranges constants somewhere or leave direct in code?

**Comment by Mygod** [comment]
Yeah I think we can add gen 7/8 pokedex number there too.

**Comment by TurtIeSocks** [comment]
> Looks good, should we make the generation ranges constants somewhere or leave direct in code?

Were you thinking that I should bring back the gen json, just with the simplified values instead of individual numbers? 

> Yeah I think we can add gen 7/8 pokedex number there too.

Will add the functions, but leaving the buttons out until we get close to the respective launches. I imagine Niantic is really going to start stretching out generations as we start to get closer and closer to the main games.

**Comment by Mygod** [comment]
No I'm saying that the button should enable both regional forms and those local Pokemon.

**Comment by versx** [comment]
> > Looks good, should we make the generation ranges constants somewhere or leave direct in code?
> 
> Were you thinking that I should bring back the gen json, just with the simplified values instead of individual numbers?
No just at the top of the file with the other variables.

**Comment by TurtIeSocks** [comment]
> No I'm saying that the button should enable both regional forms and those local Pokemon.
Hmm disagree. If Niantic keeps having Alolan & Galarian forms as special events only, then I think there's value in them being separately toggleable.

**Comment by TurtIeSocks** [comment]
@versx constants added. 

Do you have any thoughts on the Alola & Galar topic @Mygod mentioned? Either looping in regional forms with their respective generation or leaving separate.

**Comment by Mygod** [comment]
I guess given that now we have the atrocity called Kanto cup, I think filtering by national Pokedex number instead makes more sense.

File: static/js/index.js [hunk 2]
```diff
@@ -119,10 +119,19 @@ let weatherTypes = {};
...
 let skipForms = ['Shadow', 'Purified'];
 
+const kanto = [1, 151];
+const johto = [152, 251];
+const hoenn = [252, 386];
+const sinnoh = [387, 494];
+const unova = [495, 649];
+const kalos = [650, 721];
+const alola = [722, 802];
+const galar = [803, 893];
```
Comment (Mygod) [inline static/js/index.js:133]: 803-807 is Alolan ultra beasts. 808 and 809 is unknown region.

File: static/js/index.js [hunk 3]
```diff
@@ -5741,11 +5724,8 @@ function getCpAtLevel(id, form, level, isMax) {
     }
     let pkmn = [];
     if (cpMultipliers[level]) {
-        if (form === 0 || typeof masterfile.pokemon[id].forms[form].attack === 'undefined') {
-            pkmn = masterfile.pokemon[id];
-        } else {
-            pkmn = masterfile.pokemon[id].forms[form];
-        }
+        form === 0 || Object.keys(masterfile.pokemon[id].forms[form].attack).length === 0 ?
+            pkmn = masterfile.pokemon[id] : pkmn = masterfile.pokemon[id].forms[form];
```
Comment (Mygod) [inline static/js/index.js:5728]: I am not sure about operator priorities here. I recommend `pkmn = <blah> ? <stuff> : <other stuff>` instead.

File: static/js/index.js [hunk 4]
```diff
@@ -5579,85 +5585,62 @@ function getPokemonBestRank(greatLeague, ultraLeague) {
...
-            const form = pkmn.forms[formId];
-            if (form.name === 'Alola') {
-                return true;
-            }
-        }
+let masterfileFilter = (pokemonId, filter) => {
+    let matches = false;
+    const pkmn = masterfile.pokemon[pokemonId];
+    switch(filter[0]) {
+        case 'regionalForm': matches = pkmn.forms[filter[2]].name.includes(filter[1]); break;
```
Comment (Mygod) [inline static/js/index.js:5602]: I recommend not depending on `name`. Maybe it is better to use `proto` instead.

**Review by Mygod — CHANGES_REQUESTED** [review]
Event form is not included in the event filter even though I think it should.

File: static/js/index.js (hunk 3 above)
Comment (versx) [reply static/js/index.js:5728]: I agree, setting the variable based on the null coalescing operator would be much cleaner/better.
```
pkmn = form === 0 || Object.keys(masterfile.pokemon[id].forms[form].attack).length === 0
    ? masterfile.pokemon[id]
    : masterfile.pokemon[id].forms[form];
```

**Comment by versx** [comment]
> @versx constants added.
> 
> Do you have any thoughts on the Alola & Galar topic @Mygod mentioned? Either looping in regional forms with their respective generation or leaving separate.

Honestly I'm not sure what you both are mentioning lol. I think the Alola/Galar buttons should enable/disable those Pokemon.

**Comment by Mygod** [comment]
@versx We were talking about whether the Alola/Galar button should enable/disable by Pokedex or Form or both. I think now I lean towards Pokedex given the Kanto cup format.

By the way, can we take a moment to appreciate @TurtIeSocks putting his hard work to remove duplicate code in this repo during this difficult time? Well done!

File: static/js/index.js [hunk 5]
```diff
@@ -5741,11 +5727,8 @@ function getCpAtLevel(id, form, level, isMax) {
     }
     let pkmn = [];
     if (cpMultipliers[level]) {
-        if (form === 0 || typeof masterfile.pokemon[id].forms[form].attack === 'undefined') {
-            pkmn = masterfile.pokemon[id];
-        } else {
-            pkmn = masterfile.pokemon[id].forms[form];
-        }
+        pkmn = form === 0 || Object.keys(masterfile.pokemon[id].forms[form].attack).length === 0 
+        ? masterfile.pokemon[id] : masterfile.pokemon[id].forms[form];
```
Comment (Mygod) [inline static/js/index.js:5731]: Indentation.

**Review by Mygod — APPROVED** [review]
Good work!

**Review by versx — APPROVED** [review]

**Pull request merged**